File: problem48.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    void rotate(vector<vector<int>> &matrix)
    {

        // Step 1: Transpose the matrix
        for (int i = 0; i < matrix.size(); i++)
        {
            for (int j = i + 1; j < matrix.size(); j++)
            {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;
            }
        }

        // Step 2: reverse row
        // Step 2: Reverse each row
        for (int i = 0; i < matrix.size(); ++i)
        {
            reverse(matrix[i].begin(), matrix[i].end());
        }
    }
};

File: problem48.py
class Solution:
    def rotate(self, matrix: List[List[int]]):
        """
        Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        # loop from col 
        # Change one number each time 
        # Switch each number from two sides acording to the diagnal
        x = 0
        y = 0
        while x < n or y < n:
            # First add one on each x and y coordinates
            # Then increament x and y by one. 
            # If it is more then half, decrement by one
            for inc in range(x+1, n):
                temp = matrix[inc][y]
                matrix[inc][y] = matrix[x][inc]
                matrix[x][inc] = temp
            x += 1
            y += 1
        
        # Step2: switch rows of two side until them meet
        leftCol = 0
        rightCol = n-1
        while leftCol < rightCol:
            for i in range(n):
                temp = matrix[i][leftCol]
                matrix[i][leftCol] = matrix[i][rightCol]
                matrix[i][rightCol] = temp
            leftCol += 1
            rightCol -=1 
        